src/crypto/qat/qcccrypto.cc

```diff
@@ -322,150 +327,279 @@ bool QccCrypto::perform_op(unsigned char* out, const unsigned char* in,
+  Cpa32U one_batch_size = chunk_size * MAX_NUM_SYM_REQ_BATCH;
+  for (Cpa32U off = 0; off < total_len; off += one_batch_size) {
+    CompletionHandle complete;
+    for (Cpa32U offset = off, i = 0; offset < total_len && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
+      CpaCySymDpOpData *pOpData = qcc_op_mem[avail_inst].sym_op_data[i];
+      Cpa8U *pSrcBuffer = qcc_op_mem[avail_inst].src_buff[i];
+      Cpa8U *pIvBuffer = qcc_op_mem[avail_inst].iv_buff[i];
+      Cpa32U process_size = offset + chunk_size <= total_len ? chunk_size : total_len - offset;
+      if (CPA_STATUS_SUCCESS == status) {
+        // copy source into buffer
+        memcpy(pSrcBuffer, pSrc + offset, process_size);
+        // copy IV into buffer
+        memcpy(pIvBuffer, &pIv[i * ivLen], ivLen);
+      }
+      complete.add_one();
+      if (CPA_STATUS_SUCCESS == status) {
+        //pOpData assignment
+        pOpData->thisPhys = qaeVirtToPhysNUMA(pOpData);
+        pOpData->instanceHandle = qcc_inst->cy_inst_handles[avail_inst];
+        pOpData->sessionCtx = sessionCtx;
+        pOpData->pCallbackTag = (void *)&complete;
+        pOpData->cryptoStartSrcOffsetInBytes = 0;
+        pOpData->messageLenToCipherInBytes = process_size;
+        pOpData->iv = qaeVirtToPhysNUMA(pIvBuffer);
+        pOpData->pIv = pIvBuffer;
+        pOpData->ivLenInBytes = ivLen;
+        pOpData->srcBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->srcBufferLen = process_size;
+        pOpData->dstBuffer = qaeVirtToPhysNUMA(pSrcBuffer);
+        pOpData->dstBufferLen = process_size;
+      }
+
+      if (CPA_STATUS_SUCCESS == status) {
+        do {
+          status = cpaCySymDpEnqueueOp(pOpData, CPA_FALSE);
+        } while (CPA_STATUS_RETRY == status);
+        if (CPA_STATUS_SUCCESS != status) {
+          dout(1) << "cpaCySymDpEnqueueOp failed. (status = " << status << ")" << dendl;
+        }
+      }
+    }
 
-  //Copy data back to out buffer
-  memcpy(out, qcc_op_mem[avail_inst].src_buff, size);
-  //Always cleanup memory holding user-data at the end
-  memset(qcc_op_mem[avail_inst].iv_buff, 0, AES_256_IV_LEN);
-  memset(qcc_op_mem[avail_inst].src_buff, 0, qcc_op_mem[avail_inst].buff_size);
+    do {
+      status = cpaCySymDpPerformOpNow(qcc_inst->cy_inst_handles[avail_inst]);
+    } while (CPA_STATUS_RETRY == status);
+    if (CPA_STATUS_SUCCESS != status) {
+      dout(1) << "cpaCySymDpPerformOpNow failed. (status = " << status << ")" << dendl;
+    }
 
-  return true;
-}
+    if (CPA_STATUS_SUCCESS == status) {
+      complete.wait();
+    }
+
+    // Copy data back to pDst buffer
+    for (Cpa32U offset = 0, i = 0; offset < total_len && i < MAX_NUM_SYM_REQ_BATCH; offset += chunk_size, i++) {
+      Cpa8U *pSrcBuffer = qcc_op_mem[avail_inst].src_buff[i];
+      Cpa32U process_size = offset + chunk_size <= total_len ? chunk_size : total_len - offset;
+      memcpy(pDst + offset, pSrcBuffer, process_size);
+      memset(pSrcBuffer, 0, process_size);
```

Comment:
don't clear data every chunk. We can only do once for mulit chunk case.

Reply:
done.